batch/batch/worker.py

```diff
@@ -81,30 +84,94 @@ def free(self, port):
+                else:
+                    raise
+            except aiohttp.client_exceptions.ServerDisconnectedError:
+                log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
+            except asyncio.TimeoutError:
+                local_timeout *= 2
+                log.exception(f'in docker call to {f.__name__} for {name}, retrying with timeout {local_timeout} seconds', stack_info=True)
+            # exponentially back off, up to (expected) max of 30s
+            t = delay * random.random()
+            await asyncio.sleep(t)
+            delay = min(delay * 2, 60.0)
+    return _wrapper
+
+
+async def create_container(config, name):
+    error = 0
+    delay = 1
     while True:
         try:
-            return await f(*args, **kwargs)
+            return await docker.containers.create(config, name=name)
         except DockerError as e:
-            # 408 request timeout, 503 service unavailable
-            if e.status == 408 or e.status == 503:
-                log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
-            # DockerError(500, 'Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
-            # DockerError(500, 'error creating overlay mount to /var/lib/docker/overlay2/545a1337742e0292d9ed197b06fe900146c85ab06e468843cd0461c3f34df50d/merged: device or resource busy'
-            elif e.status == 500 and ("request canceled while waiting for connection" in e.message or
-                                      re.match("error creating overlay mount.*device or resource busy", e.message)):
-                log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
-            else:
+            # exponentially back off, up to (expected) max of 30s
+            t = delay * random.random()
+            await asyncio.sleep(t)
+            delay = min(delay * 2, 60.0)
+
+            error += 1
+            if error == 5:
+                log.exception(f'encountered 5 errors while creating container {name}, aborting', stack_info=True)
                 raise
-        except aiohttp.client_exceptions.ServerDisconnectedError:
-            log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
-        except asyncio.TimeoutError:
-            log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
-        # exponentially back off, up to (expected) max of 30s
-        t = delay * random.random()
-        await asyncio.sleep(t)
-        delay = min(delay * 2, 60.0)
+
+            # 409 container with name already exists
+            if e.status == 409:
+                try:
+                    return await docker.containers.get(name)
+                except DockerError as eget:
+                    # 404 No such container
+                    if eget.status == 404:
+                        continue
```

Comment:
the sleep should probably be after the `try-except`, which avoids sleeping before raising anyway and ensures we always sleep in case we use continue as we do here.

Reply:
I don't follow this comment. If the sleep is after the try-except, then we'll either continue (and skip the sleep) or we'll raise.

Reply:
I think you're overloading the meaning of the sleep code. We want to exponentially back off on issuing calls to Docker. I think, as written, we will immediately try to create if `get` returns 404. Spearately, if we hit 5 errors and the error is not 409, we'll sleep before re-raising the exception. There's no reason to sleep in that case.

I think we want something like this:
```python
async def create_container(config, name):
     errors = 0
     delay = 1
     while True:
          try:
               return await docker.containers.create(config, name=name)
          except DockerError as e:
               if e.status == 409:  # known docker issue, container might end up created anyway
                    try:
                         delay = await hailtop.utils.sleep_and_backoff(delay)
                         return await docker.containers.get(name)
                    except DockerError as eget:
                         if eget.status == 404:
                              errors += 1
                              if errors < 5:
                                   delay = await hailtop.utils.sleep_and_backoff(delay)
                                   continue
                              log.exception(f'encountered 5 errors while creating container {name}, aborting', stack_info=True)
                              raise
                         raise
               raise
```